### PR TITLE
GH-41410: [C++][FS][Azure][Docs] Add AzureFileSystem to Filesystems API reference

### PR DESCRIPTION
**Pull request by amoeba**
### Rationale for this change

See https://github.com/apache/arrow/issues/41410.

### What changes are included in this PR?

Just changes to filesystem.rst.

### Are these changes tested?

Yes, locally.

### Are there any user-facing changes?

These are those changes.
* GitHub Issue: #41410

### CODE REVIEW

**Comment by amoeba** [comment]
@github-actions crossbow submit preview-docs

**Comment by amoeba** [comment]
@github-actions crossbow submit preview-docs

**Comment by amoeba** [comment]
Preview at http://crossbow.voltrondata.com/pr_docs/41411/cpp/api/filesystem.html#azure-filesystem.

**Comment by kou** [comment]
Thanks!

Could you fix the "Supported formats" part at http://crossbow.voltrondata.com/pr_docs/41411/cpp/api/filesystem.html#_CPPv4N5arrow2fs12AzureOptions7FromUriERK3UriPNSt6stringE ?

* "i., ii., ..." style and "1., 2., ..." style are mixed (Can we control `<ol>` style?)
* Can we remove needless `\` from `...\<account>...` in `ii.`?

File: cpp/src/arrow/filesystem/azurefs.h [hunk 1]
```diff
@@ -226,7 +226,7 @@ struct ARROW_EXPORT AzureOptions {
 ///   overwriting.
 /// - When you use the ListBlobs operation without specifying a delimiter, the results
 ///   include both directories and blobs. If you choose to use a delimiter, use only a
-///   forward slash (/) -- the only supported delimiter.
+///   forward slash (/) \--- the only supported delimiter.
```
Comment (felipecrv) [inline cpp/src/arrow/filesystem/azurefs.h:229]: What is the `\-` for? I assumed `---` become an em-dash when I wrote this.

**Review by felipecrv — APPROVED** [review]

**Comment by amoeba** [comment]
@github-actions crossbow submit preview-docs

**Comment by amoeba** [comment]
Good eye @kou. Re:

> "i., ii., ..." style and "1., 2., ..." style are mixed (Can we control `<ol>` style?)

It doesn't look like it. The Doxygen output is what I expect but it looks like Breathe is getting an incorrect nesting level. [Their code](https://github.com/breathe-doc/breathe/blob/542ae9b9ed6e5bd5e92ba2360b094df8931ba756/breathe/renderer/sphinxrenderer.py#L1811-L1812) is pretty easy to follow except for how the nesting level ends up higher than what we expect. Their docs give some examples and, if you look at the final one in https://breathe.readthedocs.io/en/latest/lists.html, it actually looks wrong to me. So maybe this is by design? I don't see any config options we can use to control this.

I opted to refer to list items as (1), (2), etc instead of (i), (ii), etc so that the comment is correct on its own and accept that the rendered docs are quite right.

> Can we remove needless \ from ...\<account>... in ii.?

Done. I ended up having to escape the `@` which is unfortunate but Doxygen is handling it strangely, see my commit message.

**Comment by amoeba** [comment]
The relevant portion of the rendered docs looks like this for me now:

image

**Review by kou — APPROVED** [review]
+1

Could you fix lint error?

**Comment by amoeba** [comment]
Pushed an attempt but it looks like we might have to get creative to satisfy the linter. I'll return to this in a few hours. I originally smushed list items into their own line because they weren't rendering but there's probably a way to prevent a source newline from resulting in a break in the HTML.

**Comment by kou** [comment]
@github-actions crossbow submit preview-docs

**Comment by amoeba** [comment]
I ended up reverting the changes which fixes the output to appease the linter here. This is how things look now:

Screenshot 2024-04-30 at 4 34 25 PM

I'm not sure of how to get Doxygen to produce the output I want here so maybe we just merge this as-is.

**Pull request merged**